File: sample-objc/ARGEARSample/UI/UIComponent/Main/Bottom/Filter/FilterCollectionView.m
//
//  FilterCollectionView.m
//  ARGEARSample
//
//  Created by Jihye on 10/09/2019.
//  Copyright © 2019 Seerslab. All rights reserved.
//

#import "FilterCollectionView.h"
#import "NetworkManager.h"
#import "FilterManager.h"

#define kFilterCellNibName @"FilterCollectionViewCell"
#define kFilterCellIdentifier @"filtercell"

@implementation FilterCollectionView

/*
// Only override drawRect: if you perform custom drawing.
// An empty implementation adversely affects performance during animation.
- (void)drawRect:(CGRect)rect {
    // Drawing code
}
*/

- (void)awakeFromNib {
    [super awakeFromNib];
    // Initialization code
    
    [self setDelegate:self];
    [self setDataSource:self];
    
    [self registerNib:[UINib nibWithNibName:kFilterCellNibName bundle:nil] forCellWithReuseIdentifier:kFilterCellIdentifier];
    
    _selectedIndexPath = nil;
}

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath {
    
    if ([_selectedIndexPath isEqual:indexPath]) {
        return;
    }
    
    Item *item = _filtersArray[[indexPath row]];
    
    __weak FilterCollectionView *weakSelf = self;
    [[FilterManager shared] setFilter:item success:^{
        [weakSelf setSelectedIndexPath:indexPath];

        [collectionView reloadData];
    } fail:^{
    }];
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    return [_filtersArray count];
}

- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView {
    return 1;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    FilterCollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:kFilterCellIdentifier forIndexPath:indexPath];

    [cell unchecked];
    if ([indexPath isEqual:_selectedIndexPath]) {
        [cell checked];
    }
    
    [cell setData:_filtersArray[[indexPath row]]];
    [cell setFilterNameLabelColor:[self tag]];
    
    return cell;
}

- (void)setRatio:(ARGMediaRatio)ratio {
    
    [self setTag:ratio];
    [self reloadData];
}

@end

File: sample-swift/ARGearSample/UI/UIComponent/Main/Bottom/Filter/FilterCollectionView.swift
//
//  FilterCollectionView.swift
//  ARGearSample
//
//  Created by Jihye on 2020/01/30.
//  Copyright © 2020 Seerslab. All rights reserved.
//

import UIKit
import ARGear

class FilterCollectionView: ARGBottomFunctionBaseCollectionView {
    
    let kFilterCellNibName = "FilterCollectionViewCell"
    let kFilterCellIdentifier = "filtercell"
    
    var filters: [Item]? {
        didSet {
            guard let _ = filters else {
                return
            }
            DispatchQueue.main.async {
                self.reloadData()
            }
        }
    }
    
    override func awakeFromNib() {
        super.awakeFromNib()
        
        self.delegate = self
        self.dataSource = self
        
        self.register(UINib(nibName: kFilterCellNibName, bundle: nil), forCellWithReuseIdentifier: kFilterCellIdentifier)
    }

}

extension FilterCollectionView: UICollectionViewDelegate {
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        if let selectedIndexPath = self.selectedIndexPath {
            if selectedIndexPath == indexPath {
                return
            }
        }
        
        if indexPath.row == 0 {
            FilterManager.shared.clearFilter()
            return
        }
        
        if let filters = filters {
            let filter = filters[indexPath.row - 1]
            
            FilterManager.shared.setFilter(filter, successBlock: { [weak self] in
                guard let self = self else { return }
                
                DispatchQueue.main.async {
                    self.selectedIndexPath = indexPath
                    self.reloadData()
                }
            }) {
            }
        }
    }
}

extension FilterCollectionView: UICollectionViewDataSource {
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        if let filters = self.filters {
            return filters.count + 1
        } else {
            return 0
        }
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let filterCell = collectionView.dequeueReusableCell(withReuseIdentifier: kFilterCellIdentifier, for: indexPath) as! FilterCollectionViewCell
        
        if indexPath.row == 0 {
            filterCell.disableButtonCell(ratio: ARGMediaRatio(rawValue: self.tag)!)
            return filterCell
        }
        
        filterCell.unchecked()
        if indexPath == self.selectedIndexPath {
            filterCell.checked()
        }
        
        if let filters = self.filters {
            filterCell.setFilter(filters[indexPath.row - 1])
            filterCell.setRatio(ARGMediaRatio(rawValue: self.tag)!)
        }
        return filterCell
    }
}
